test_conformance/spirv_new/types.hpp

```diff
@@ -0,0 +1,215 @@
+{
+    T0 val0;
+    T1 val1;
+};
+
+
+template<typename T0, typename T1>
+inline bool operator != (const AbstractStruct2<T0, T1> &lhs,
+                         const AbstractStruct2<T0, T1> &rhs)
+{
+    return lhs.val0 != rhs.val0 || lhs.val1 != rhs.val1;
+}
+
+
+template<typename T> struct is_double { static const bool value = false; };
+template<> struct  is_double<cl_double> { static const bool value = true; };
+template<> struct  is_double<cl_double2> { static const bool value = true; };
+
+template<typename T>
+T genrandReal(RandomSeed &seed)
+{
+    return genrand_real1(seed);
+}
+
+template<typename T, int N>
+T genrandRealVec(RandomSeed &seed)
+{
+    T res;
+    for (int i = 0; i < N; i++) {
+        res.s[i] = genrand_real1(seed);
+    }
+    return res;
+}
+
+#define GENRAND_REAL_FUNC(TYPE, N)                                      \
+    template<> inline TYPE##N genrandReal<TYPE##N>(RandomSeed &seed)    \
+    {                                                                   \
+        return genrandRealVec<TYPE##N, N>(seed);                        \
+    }                                                                   \
+
+GENRAND_REAL_FUNC(cl_float, 2)
+GENRAND_REAL_FUNC(cl_float, 4)
+GENRAND_REAL_FUNC(cl_double, 2)
+GENRAND_REAL_FUNC(cl_double, 4)
+
+template<> inline cl_half genrandReal<cl_half>(RandomSeed &seed)
+{
+    return (cl_half)(genrand_int32(seed) % 2048);
+}
+
+template<typename T>
+T genrand(RandomSeed &seed)
+{
+    return genrandReal<T>(seed);
+}
+
+template<> inline cl_int genrand<cl_int>(RandomSeed &seed)
+{
+    return genrand_int32(seed);
+}
+
+template<> inline cl_long genrand<cl_long>(RandomSeed &seed)
+{
+    return genrand_int32(seed);
+}
+
+template<> inline cl_short genrand<cl_short>(RandomSeed &seed)
+{
+    return genrand_int32(seed);
+}
+
+#define GENRAND_INT_VEC(T, N)                               \
+    template<> inline T##N genrand<T##N>(RandomSeed &seed)  \
+    {                                                       \
+        T##N res;                                           \
+        for (int i = 0; i < N; i++) {                       \
+            res.s[i] = (T)genrand_int32(seed);              \
+        }                                                   \
+        return res;                                         \
+    }                                                       \
+
+GENRAND_INT_VEC(cl_int, 4)
+GENRAND_INT_VEC(cl_uint, 4)
+GENRAND_INT_VEC(cl_long, 2)
+GENRAND_INT_VEC(cl_char, 16)
+
+template<typename Tv>
+Tv negOp(Tv in)
+{
+    return -in;
+}
+
+template<typename Tv>
+Tv notOp(Tv in)
+{
+    return ~in;
+}
+
+template<typename Tv, int N>
+Tv negOpVec(Tv in)
+{
+    Tv out;
+    for (int i = 0; i < N; i++) {
+        out.s[i] = -in.s[i];
+    }
+    return out;
+}
+
+template<typename Tv, int N>
+Tv notOpVec(Tv in)
+{
+    Tv out;
+    for (int i = 0; i < N; i++) {
+        out.s[i] = ~in.s[i];
+    }
+    return out;
+}
+
+template <typename T>
```

Comment:
this definition and changes to use of `vector` in a lot of the test files are from #177